File: ui/src/__tests__/factories.ts
import type {
  Changeset,
  ChangesetSummary,
  LinkTargetInfo,
  ProposedChange,
  ContentItem,
  PassageAnnotation,
  TagInfo,
  VaultTaxonomy,
  ZoteroPaperSummary,
  ZoteroAnnotationItem,
  ZoteroCollection,
} from "../types";

export function makeContentItem(overrides?: Partial<ContentItem>): ContentItem {
  return {
    text: "Neural networks are universal approximators.",
    source: "https://example.com/article",
    source_type: "web",
    ...overrides,
  };
}

export function makeProposedChange(
  overrides?: Partial<ProposedChange>,
): ProposedChange {
  return {
    id: "change-1",
    tool_name: "create_note",
    input: { path: "Papers/Test.md", content: "# Test" },
    original_content: null,
    proposed_content: "# Test\n\nContent.",
    diff: "--- a/Papers/Test.md\n+++ b/Papers/Test.md\n@@ -0,0 +1,3 @@\n+# Test\n+\n+Content.\n",
    status: "pending",
    ...overrides,
  };
}

export function makeChangeset(overrides?: Partial<Changeset>): Changeset {
  return {
    id: "cs-test-1",
    items: [makeContentItem()],
    changes: [makeProposedChange()],
    reasoning: "Created a new note.",
    status: "pending",
    created_at: "2024-01-01T00:00:00Z",
    source_type: "web",
    routing: {
      action: "create",
      target_path: "Papers/Test.md",
      reasoning: "New topic.",
      confidence: 0.9,
      additional_targets: null,
      duplicate_notes: null,
    },
    usage: null,
    feedback: null,
    parent_changeset_id: null,
    ...overrides,
  };
}

export function makeChangesetSummary(
  overrides?: Partial<ChangesetSummary>,
): ChangesetSummary {
  return {
    id: "cs-test-1",
    status: "pending",
    created_at: "2024-01-01T00:00:00Z",
    source_type: "web",
    change_count: 1,
    routing: {
      action: "create",
      target_path: "Papers/Test.md",
      reasoning: "New topic.",
      confidence: 0.9,
      additional_targets: null,
      duplicate_notes: null,
    },
    feedback: null,
    parent_changeset_id: null,
    ...overrides,
  };
}

export function makePaper(
  overrides?: Partial<ZoteroPaperSummary>,
): ZoteroPaperSummary {
  return {
    key: "PAPER1",
    title: "Test Paper Title",
    authors: ["Smith, John", "Doe, Jane"],
    year: "2024",
    item_type: "journalArticle",
    last_synced: null,
    changeset_id: null,
    annotation_count: 5,
    ...overrides,
  };
}

export function makeAnnotation(
  overrides?: Partial<ZoteroAnnotationItem>,
): ZoteroAnnotationItem {
  return {
    key: "ANN1",
    text: "Important finding about neural networks.",
    comment: "Key result",
    color: "#ffd400",
    page_label: "42",
    annotation_type: "highlight",
    date_added: "2024-01-01T00:00:00Z",
    ...overrides,
  };
}

export function makePassageAnnotation(
  overrides?: Partial<PassageAnnotation>,
): PassageAnnotation {
  return {
    id: "ann-1",
    selectedText: "Neural networks are universal approximators.",
    comment: "Clarify which theorem this refers to",
    ...overrides,
  };
}

export function makeCollection(
  overrides?: Partial<ZoteroCollection>,
): ZoteroCollection {
  return {
    key: "COL1",
    name: "My Collection",
    parent_collection: null,
    num_items: 5,
    num_collections: 0,
    ...overrides,
  };
}

export function makeTagInfo(overrides?: Partial<TagInfo>): TagInfo {
  return { name: "research", count: 10, ...overrides };
}

export function makeLinkTargetInfo(
  overrides?: Partial<LinkTargetInfo>,
): LinkTargetInfo {
  return { title: "Machine Learning", count: 5, ...overrides };
}

export function makeVaultTaxonomy(
  overrides?: Partial<VaultTaxonomy>,
): VaultTaxonomy {
  return {
    folders: ["Papers", "Topics", "Projects", "daily"],
    tags: [
      makeTagInfo({ name: "research", count: 15 }),
      makeTagInfo({ name: "research/ai", count: 8 }),
      makeTagInfo({ name: "research/ml", count: 5 }),
      makeTagInfo({ name: "paper", count: 20 }),
      makeTagInfo({ name: "daily", count: 45 }),
    ],
    tag_hierarchy: [
      { name: "daily", children: [], description: null },
      { name: "paper", children: [], description: null },
      {
        name: "research",
        children: [
          { name: "ai", children: [], description: null },
          { name: "ml", children: [], description: null },
        ],
        description: null,
      },
    ],
    link_targets: [
      makeLinkTargetInfo({ title: "Machine Learning", count: 12 }),
      makeLinkTargetInfo({ title: "Projects/My Project", count: 3 }),
    ],
    total_notes: 142,
    ...overrides,
  };
}

File: tests/factories.py
"""Test data factories for building valid model instances with sensible defaults."""

import uuid
from datetime import datetime, timezone

from src.models import (
    Changeset,
    ContentItem,
    ProposedChange,
    RoutingInfo,
    SourceMetadata,
)


def make_content_item(**overrides) -> ContentItem:
    defaults = {
        "text": "This is a highlighted passage about neural networks.",
        "source": "https://example.com/article",
        "annotation": None,
        "source_type": "web",
        "color": None,
        "source_metadata": None,
    }
    defaults.update(overrides)
    return ContentItem(**defaults)


def make_zotero_content_item(**overrides) -> ContentItem:
    defaults = {
        "text": "The model achieves state-of-the-art results on ImageNet.",
        "source": "Deep Learning Paper",
        "annotation": "Key result",
        "source_type": "zotero",
        "color": "#ffd400",
        "source_metadata": SourceMetadata(
            title="Deep Learning for Computer Vision",
            doi="10.1234/test.2024",
            authors=["Smith, John", "Doe, Jane"],
            year="2024",
            paper_key="ABC123",
        ),
    }
    defaults.update(overrides)
    return ContentItem(**defaults)


def make_proposed_change(**overrides) -> ProposedChange:
    defaults = {
        "id": str(uuid.uuid4()),
        "tool_name": "create_note",
        "input": {"path": "Papers/Test Note.md", "content": "# Test\n\nContent."},
        "original_content": None,
        "proposed_content": "# Test\n\nContent.",
        "diff": "--- a/Papers/Test Note.md\n+++ b/Papers/Test Note.md\n@@ -0,0 +1,3 @@\n+# Test\n+\n+Content.\n",
        "status": "pending",
    }
    defaults.update(overrides)
    return ProposedChange(**defaults)


def make_routing_info(**overrides) -> RoutingInfo:
    defaults = {
        "action": "create",
        "target_path": "Papers/Test Note.md",
        "reasoning": "No existing note covers this topic.",
        "confidence": 0.9,
    }
    defaults.update(overrides)
    return RoutingInfo(**defaults)


def make_changeset(**overrides) -> Changeset:
    defaults = {
        "id": str(uuid.uuid4()),
        "items": [make_content_item()],
        "changes": [make_proposed_change()],
        "reasoning": "Created a new note for this content.",
        "status": "pending",
        "created_at": datetime.now(timezone.utc).isoformat(),
        "source_type": "web",
        "routing": make_routing_info(),
    }
    defaults.update(overrides)
    if defaults["items"] is None:
        del defaults["items"]
    return Changeset(**defaults)


def make_replace_change(**overrides) -> ProposedChange:
    defaults = {
        "id": str(uuid.uuid4()),
        "tool_name": "replace_note",
        "input": {"path": "Notes/Test.md", "content": "# Updated\n\nNew content."},
        "original_content": "# Test\n\nOld content.",
        "proposed_content": "# Updated\n\nNew content.",
        "diff": "--- a/Notes/Test.md\n+++ b/Notes/Test.md\n@@ -1,3 +1,3 @@\n-# Test\n+# Updated\n \n-Old content.\n+New content.\n",
        "status": "pending",
    }
    defaults.update(overrides)
    return ProposedChange(**defaults)


def make_delete_change(**overrides) -> ProposedChange:
    defaults = {
        "id": str(uuid.uuid4()),
        "tool_name": "delete_note",
        "input": {"path": "Notes/Obsolete.md"},
        "original_content": "# Obsolete\n\nOld content.",
        "proposed_content": "",
        "diff": "--- a/Notes/Obsolete.md\n+++ b/Notes/Obsolete.md\n@@ -1,3 +0,0 @@\n-# Obsolete\n-\n-Old content.\n",
        "status": "pending",
    }
    defaults.update(overrides)
    return ProposedChange(**defaults)
